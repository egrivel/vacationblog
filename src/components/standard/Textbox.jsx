'use strict';

import React from 'react';
import createReactClass from 'create-react-class';
import PropTypes from 'prop-types';

import Formrow from './Formrow';

const Textbox = createReactClass({
  displayName: 'Textbox',

  propTypes: {
    fieldId: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    value: PropTypes.string,
    onChange: PropTypes.func.isRequired
  },

  _onChange: function(event) {
    this.props.onChange(event.target.value, this.props.fieldId);
  },

  render: function() {
    let value = this.props.value;
    if (!value) {
      value = '';
    }

    return (
      <Formrow
        key={'k-' + this.props.fieldId}
        label={this.props.label}
        labelFor={this.props.fieldId}
      >
        <input
          id={this.props.fieldId}
          type="text"
          value={value}
          onChange={this._onChange}
        />
      </Formrow>
    );
  }
});

export default Textbox;
