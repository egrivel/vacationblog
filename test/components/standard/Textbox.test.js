import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import enzyme from 'enzyme';

import Textbox from '../../../src/components/standard/Textbox';

/*
 * Test strategy
 * The component has one optional prop (value) and one callback
 * function (onChange). The requirements are:
 *  - the component renders with or without a value
 *  - the onChange is fired when changing the value
 */
describe('components/standard/Textbox', () => {
  const testFieldId = 'test-field-id';
  const testLabel = 'Test Label';
  const testValue = 'This is the test value';
  const newTestValue = 'This is the changed test value';
  let onChangeStub;
  let testEvent;

  beforeEach(() => {
    onChangeStub = sinon.stub();
    testEvent = {
      preventDefault() {},
      target: {}
    };
  });

  afterEach(() => {
    sinon.restore();
  });

  describe('#render', () => {
    test('renders without a value', () => {
      const render = renderer.create(
        <Textbox
          fieldId={testFieldId}
          label={testLabel}
          onChange={onChangeStub}
        />
      );
      expect(render.toJSON()).toMatchSnapshot();
    });

    test('renders with a value', () => {
      const render = renderer.create(
        <Textbox
          fieldId={testFieldId}
          label={testLabel}
          onChange={onChangeStub}
          value={newTestValue}
        />
      );
      expect(render.toJSON()).toMatchSnapshot();
    });
  });

  describe('handles onChange', () => {
    it('passes the new value back', () => {
      const component = enzyme.shallow(
        <Textbox
          fieldId={testFieldId}
          label={testLabel}
          onChange={onChangeStub}
          value={testValue}
        />
      );
      testEvent.target.value = newTestValue;
      component.find('input').simulate('change', testEvent);
      expect(onChangeStub.callCount).toBe(1);
      expect(onChangeStub.args[0][0]).toBe(newTestValue);
      expect(onChangeStub.args[0][1]).toBe(testFieldId);
    });
  });
});
