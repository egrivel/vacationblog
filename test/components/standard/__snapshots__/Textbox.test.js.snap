// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/standard/Textbox #render renders with a value 1`] = `
<div>
  <div
    className="formLabel"
    htmlFor="test-field-id"
  >
    Test Label
  </div>
  <div
    className="formValue"
  >
    <input
      id="test-field-id"
      onChange={[Function]}
      type="text"
      value="This is the changed test value"
    />
  </div>
</div>
`;

exports[`components/standard/Textbox #render renders without a value 1`] = `
<div>
  <div
    className="formLabel"
    htmlFor="test-field-id"
  >
    Test Label
  </div>
  <div
    className="formValue"
  >
    <input
      id="test-field-id"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
</div>
`;
